### PR TITLE
Autobuild: Refactor & use actions/cache

### LINKED ISSUE #2277: Build: Use actions/cache
**Describe the bug**

We should use actions/cache to improve build speeds and resiliency against temporary flakiness on external dependencies.

The following things to check come to mind:
- [x] aqt: Can keep downloaded archives, but won't re-use them. Caching the whole Qt directories for that reason now.
- [x] choco: Hardcoding a `cacheLocation` now
- [x] brew: not used
- [x] pip: Different paths for Windows/Mac
- [x] Android SDK & NDK
- [ ] brew if added via #2207

Depends on #2276 to avoid caching a bad state.

**To Reproduce**

Random flakiness such as Qt download failures.

**Expected behavior**

Dependencies should be cached. Temporary download failures should not break the build.

**Screenshots**

https://github.com/jamulussoftware/jamulus/runs/4936344409?check_suite_focus=true
```
Caught ArchiveConnectionError, terminating installer workers
Failure to connect to https://download.qt.io/online/qtsdkrepository/windows_x86/desktop/qt5_5152/qt.qt5.5152.win64_msvc2019_64/5.15.2-0-202011130602qtremoteobjects-Windows-Windows_10-MSVC2019-Windows-Windows_10-X86_64.7z.sha1: ReadTimeout
```

**Additional context**

https://docs.github.com/en/actions/advanced-guides/caching-dependencies-to-speed-up-workflows
https://github.com/actions/cache

### PR DESCRIPTION
**Pull request by hoffie**
**Short description of changes**

This PR introduces an actions/cache step for Android, Mac and Windows builds. For Linux, this makes little sense as all dependencies are installed via apt, which should be almost locally anyway (Azure apt mirror).
    
For the caching to be effective and correct, all cached versions must be pinned. This is permits the caching logic to invalidate caches automatically based on a key (script content + invocation paramters). Pinning has partly been implemented already (Qt, NSIS, ASIOSDK, Android Command Line Tools, Android NDK, Android SDK). In addition, this PR requires #2345 which introduces version pinning for JACK and aqtinstaller.
    
The following environments are cached:
- Qt installation directory (Windows, Mac, Android)
- pip cache (Windows, Mac)
- choco cache (Windows)
- Android SDK & NDK (Android)
    
The cache is used if the workflow-provided calls (which include Qt versions) are unchanged and if certain files (most notably autobuild.yml, prepare scripts and prepare script dependencies) are unchanged. If anything changes, the cache is not used. Instead, everything starts clean and the result is cached for further reuse again.
    
Goals of this PR:
- Increase resilience against temporary external outages (e.g. Qt mirror problems)
- Make builds more deterministic
- Speed up builds by avoiding repeated downloads and some extractions

Risks:
- If we get our cache key wrong we might end up building in an unexpected build environment state. I took care to ensure that we fail in such cases, but there is no guarantee. We should check the cache keys extensively. Cache invalidation is hard (on the other hand: The [actions/cache](https://github.com/actions/cache) seems to be wildly used).
- Github Actions caches are limited to 10GB. If we reach this limit, old cache keys are evicted automatically and we no longer profit from caching. A full autobuild run corrently takes about 2300 MB (with Android taking the vast amount of it). Unless there are lots of pending PRs which touch autobuild-related code, this should cover our use case most of the time. See https://github.com/jamulussoftware/jamulus/issues/2277#issuecomment-1022014373 for further details.

Several refactorings in the build scripts have been necessary to be able to introduce those features. Therefore, they are part of this PR. In order to keep history clean and make review easier, they have been extracted as individual commits:
- Define versions and paths as variables
- Android: Avoid cleanups, condense calls
- Clean up Android Docker leftovers
- Use bash consistently

**Context: Fixes an issue?**

Fixes #2277

**Does this change need documentation? What needs to be documented and how?**

No.

**Status of this Pull Request**

**Not** ready for merge. Should wait for 3.9.0.

**What is missing until this pull request can be merged?**

- [x] #2276 should be merged first to avoid caching failed builds
- [ ] Thorough review of the Cache paths and `key` logic
- [ ] All officially supported release artifacts should be validated
  - [ ] Windows
  - [ ] Windows JACK (?)
  - [ ] Mac
  - [ ] Mac Legacy
  - [ ] Debian/Ubuntu

## Checklist

- [x] I've verified that this Pull Request follows the [general code principles](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#jamulus-projectsource-code-general-principles)
- [x] I tested my code and it does what I want
- [ ] ~My code follows the [style guide](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#source-code-consistency)~ 
- [x] I waited some time after this Pull Request was opened and all GitHub checks completed without errors. 
- [x] I've filled all the content above

### CODE REVIEW

File: .github/workflows/autobuild.yml [hunk 1]
```diff
@@ -165,6 +165,38 @@ jobs:
...
+      - name:                       "Cache Android dependencies"
+        if:                         ${{ matrix.config.target_os == 'android' }}
+        uses:                       actions/cache@v2
+        with:
+          path: |
+            /opt/Qt
+            /opt/android/android-sdk
+            /opt/android/android-ndk
+          key:                      ${{ matrix.config.target_os }}-${{ hashFiles('.github/workflows/autobuild.yml', 'autobuild/android/autobuild_apk_1_prepare.sh', 'autobuild/android/install-qt.sh') }}-${{ matrix.config.cmd1_prebuild }}
+
```
Comment (hoffie) [inline .github/workflows/autobuild.yml:199]: I've considered making this a single stel and providing the platform-specific differences (`path`, `key`) via the matrix config, but that would imply that we would have to duplicate all that per-instance (e.g. once for regular Windows builds, once for JACK; once for regular Mac builds, once for legacy). Therefore I decided to go this route.

**Review by pljones — APPROVED** [review]
Looks better and has lots green builds so far.

**Review by ann0see — APPROVED** [review]
Ready to merge?

**Review by softins — APPROVED** [review]
This is looking good

**Comment by hoffie** [comment]
Thanks for the reviews.
While the refactoring stuff is quite safe, getting the cache keys even slightly wrong might result in unexpected build output. Therefore I think it might be better to target that for 3.9.0 and avoid having build issues during 3.8.2 release time?
Let me know if you think otherwise.

**Comment by hoffie** [comment]
What would feel safe with would be dropping the [cache logic](https://github.com/jamulussoftware/jamulus/pull/2284/commits/e8c05a4687c692c0d6d9d463075c2f1abf847488) from this PR, merge this PR for 3.8.2 and open another PR with the caching stuff for 3.9.0.

**Comment by ann0see** [comment]
Should be merged after upcoming release (not squash merged)

**Comment by ann0see** [comment]
@hoffie are you ok with a merge (and test during our development cycle) or is anything blocking?

**Comment by hoffie** [comment]
> @hoffie are you ok with a merge (and test during our development cycle)

Yes, I think that'd be best.

> is anything blocking (i.e. caching of create-dmg, the PR I merged now)

Should not be blocking. Brew is not explicitly part of the cache in this PR, but that's not an issue. We can extend it in a follow-up. I neither want to sneak this into this already approved PR nor go another round of approvals here. Will open a follow-up.

**Pull request merged**